src/server/journal/journal_slice.cc

```diff
@@ -74,17 +74,19 @@ void JournalSlice::SetFlushMode(bool allow_flush) {
   }
 }
 
+size_t JournalSlice::GetRingBufferSize() const {
+  return ring_buffer_.size();
+}
+
 void JournalSlice::AddLogRecord(const Entry& entry) {
   DCHECK(ring_buffer_.capacity() > 0);
 
-  ring_buffer_.push_back(JournalItem());
-  JournalItem& item = ring_buffer_.back();
+  JournalItem item;
```

Comment:
What could happen here is that, `CallOnChange` is preemptive. So if for some reason we push too many items in the ring buffer we might wake up a block `CallOnChange` with its `item deallocated from the ring buffer`.

To fix this we create the item here and push it after we call `ConsumeOnJournal` but before `ThrottleIfNeeded` preserving the `lsn order` within the `ring buffer`

@adiholden wdyt ?

Reply:
Can you explain more the scenario?
CallOnChange can preempt but this is after calling ConsumeJournalChange which needs the journal item. If we preempt in ThrottleIfNeeded and the journal item is not already in the buffer there should not be any issue.